python/ray/rllib/common.py

```diff
@@ -115,17 +147,36 @@ def save(self):
             Checkpoint path that may be passed to restore().
         """
 
-        raise NotImplementedError
+        checkpoint_path = self._save()
+        pickle.dump(
+            [self.experiment_id, self.iteration, self.timesteps_total,
+             self.time_total_s],
+            open(checkpoint_path + ".rllib_metadata", "wb"))
+        return checkpoint_path
 
     def restore(self, checkpoint_path):
         """Restores training state from a given model checkpoint.
 
         These checkpoints are returned from calls to save().
         """
 
-        raise NotImplementedError
+        self._restore(checkpoint_path)
+        metadata = pickle.load(open(checkpoint_path + ".rllib_metadata", "rb"))
+        self.experiment_id = metadata[0]
+        self.iteration = metadata[1]
+        self.timesteps_total = metadata[2]
+        self.time_total_s = metadata[3]
 
     def compute_action(self, observation):
         """Computes an action using the current trained policy."""
 
         raise NotImplementedError
+
+    def _train(self):
```

Comment:
want to add some brief documentation saying what the underscore methods (_train, _save, _load) are supposed to do and that they are to be implemented by the algorithms?

Reply:
Done